cpp/src/KIM_FilesystemPath.hpp

```diff
@@ -0,0 +1,434 @@
+//
+// CDDL HEADER START
+//
+// The contents of this file are subject to the terms of the Common Development
+// and Distribution License Version 1.0 (the "License").
+//
+// You can obtain a copy of the license at
+// http://www.opensource.org/licenses/CDDL-1.0.  See the License for the
+// specific language governing permissions and limitations under the License.
+//
+// When distributing Covered Code, include this CDDL HEADER in each file and
+// include the License file in a prominent location with the name LICENSE.CDDL.
+// If applicable, add the following below this CDDL HEADER, with the fields
+// enclosed by brackets "[]" replaced with your own identifying information:
+//
+// Portions Copyright (c) [yyyy] [name of copyright owner]. All rights reserved.
+//
+// CDDL HEADER END
+//
+
+//
+// Copyright (c) 2016--2019, Regents of the University of Minnesota.
+// All rights reserved.
+//
+// Contributors:
+//    Ryan S. Elliott
+//
+
+//
+// Release: This file is part of the kim-api.git repository.
+//
+
+
+#ifndef KIM_FILESYSTEM_PATH_HPP_
+#define KIM_FILESYSTEM_PATH_HPP_
+
+#include <cstdlib>
+#include <iostream>
+
+// If available on the local platform, use the std::filesystem::path class
+// for cross-platform file path handling introduced with C++17. Otherwise,
+// fall back to our own, minimal implementation of the file path class based
+// on std::string.
+#if __cplusplus >= 201703L && defined(__has_include) && __has_include(<filesystem>)
+
+#include <filesystem>
+#include <random>
+
+namespace KIM
+{
+namespace FILESYSTEM
+{
+class Path : public std::filesystem::path
```

Comment:
I think I would prefer to keep the `path` object a direct drop-in for the C++17 standard.  This way, we can eventually remove the wrappers around standard stuff.  I suppose that means `Path` should become `path`.

In that case, what about making it a simple typedef for the windows case?

Reply:
I kind of agree with you, but it's not a clear-cut decision, I would say. The current solution is more than just a simple "typedef", even in C++17 mode. It adds some additional logic and, more importantly, it allows the legacy C++98 client code to use the new filesystem API. `std::filesystem` makes use of C++11 facilities such as `std::error_code`, which are not available to the client code. Meaning we would have to also develop stand-ins for them. 

It's a close call, but I would suggest to use the current solution until a change from C++98 to C++11 is made for the entire code base. Then it makes more sense to work towards a simple `typedef`. 

But it is your call. I can change it if you tell me to, no problem.

Reply:
I see, this is helpful.  Thanks.

What do you think about a clean header file that defines the `KIM::FILESYSTEM::Path` class without inheritance?  The implementation would go in a `.cpp` file and would need to have a member variable of the appropriate type.  Even though `Path` is an internal api, this would make it easier to lookup and understand (for me) its interface.

Reply:
Following your suggestion, I have now merged the two alternative implementations of the `Path` class into a single class definition, avoided the use of class inheritance by introducing a private field instead, and moved the implementation of most class methods into a separate `.cpp` file.

Reply:
Alex, this is great.  We're almost there.  I'll submit a new, clean review